### PR TITLE
#8621 Widgets on map mobile view

### PR DESCRIPTION
**Pull request by alex-fko**
## Description
This PR provides all the necessary changes to make map widgets work properly on mobile (as well as tablets).

- It updates grid layout to show single widget.
- It replaces widget title with an extra controls (select box and arrows) to allow user to toggle currently active widget.
- It simplifies behavior of widgets tray allowing only to collapse and expand widget; ability to toggle visibility of specific widget is hidden.
- Pinning, editing are deactivated for mobile view.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [x] Other... Please describe: enhancement

## Issue

**What is the current behavior?**
#8621 

**What is the new behavior?**
Widgets are available for mobiles and tablets

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/components/widgets/widget/MobileControls.jsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import React, { useMemo } from 'react';
+import Select from 'react-select';
+import { Glyphicon } from 'react-bootstrap';
+import classnames from 'classnames';
+
+import Button from "../../../components/misc/Button";
+
+const MobileControls = ({ options }) => {
```
Comment (allyoucanmap) [inline web/client/components/widgets/widget/MobileControls.jsx:8]: I would prefer a different name because even we are using this on mobile mainly it seems to have an effect also on desktop after shrinking the screen width ( eg. WidgetSelector, WidgetInlineSelector, ... )

File: web/client/plugins/Widgets.jsx [hunk 2]
```diff
@@ -37,13 +53,17 @@ compose(
             getMaximizedState,
             dependenciesSelector,
             (state) => mapLayoutValuesSelector(state, { right: true}),
-            (id, widgets, layouts, maximized, dependencies, mapLayout) => ({
+            state => state.browser && state.browser.mobile,
```
Comment (allyoucanmap) [inline web/client/plugins/Widgets.jsx:56]: I think we have already an isMobile selector anyway I'm wondering if we should rely only on the mobile selector, we already know that if a plugin is listed in the plugins.mobile list of localConfig this will be available only on mobile device. We could use a cfg only on mobile to enable the single widget visualization.

File: web/client/plugins/Widgets.jsx [hunk 3]
```diff
@@ -160,6 +251,7 @@ class Widgets extends React.Component {
  * @memberof plugins
  * @name Widgets
  * @class
+ * @prop {boolean} [showMobileNavigation] show arrows to toggle between available widgets in mobile view. True by default.
```
Comment (allyoucanmap) [inline web/client/plugins/Widgets.jsx:254]: Why do we need an option to disable/enable only the arrows?

File: web/client/themes/default/less/widget.less [hunk 4]
```diff
@@ -154,6 +154,57 @@
...
+            }
+
+            .previous-widget {
+                left: calc(25% - 15px);
+                transform: translateX(-50%);
+            }
+            .next-widget {
+                right: calc(25% - 15px);
+                transform: translateX(+50%);
+            }
```
Comment (allyoucanmap) [inline web/client/themes/default/less/widget.less:205]: Why do we need to move this two buttons? I think it will be better to let them follow the flex layout default behavior

File: web/client/themes/default/less/widget.less [hunk 5]
```diff
@@ -154,6 +154,57 @@
...
+                        }
+                    }
+                    .Select-input {
+                        height: 24px;
+                    }
+                }
+                .Select-menu-outer {
+                    margin-top: 2px;
+                }
+            }
```
Comment (allyoucanmap) [inline web/client/themes/default/less/widget.less:191]: Do we need all these changes on the select component?

**Comment by alex-fko** [comment]
@allyoucanmap It's not interfering with desktop layout. `state.browser.mobile` is not affected by window width change, it relies on browser user agent. Therefore `isTablet` and `isMobile` are additional checks to precisely determine whether mobile browser is running on tablet or mobile.
Layout behave exactly the same way on dev, switching between multi column and single column, it's probably somehow related to the width checks but I'm not sure if changing this is in a scope of a task.
both `isTablet` and `isMobile` are needed just to make widgets usage experience better on tablets where it's slightly more free space to show widgets, therefore they can be displayed in desktop-like mode with some limitations (e.g. no edit mode & limited grid width etc.).

File: web/client/components/widgets/widget/MobileControls.jsx (hunk 1 above)
Comment (alex-fko) [reply web/client/components/widgets/widget/MobileControls.jsx:8]: It sounds reasonable to change the name, I like WidgetInlineSelector, but it doesn't effect desktop, otherwise select list would be visible there on shrinking browser window.

File: web/client/plugins/Widgets.jsx (hunk 2 above)
Comment (alex-fko) [reply web/client/plugins/Widgets.jsx:56]: This is an option, but even in this case we need to distinguish between Mobiles where viewport is usually below 600px (400+px actually) and tablets where it's more than 600px (checked here https://screensiz.es/tablet). For tablets it makes sense to have several columns because there are more space to utilize by widgets, especially in landscape mode

File: web/client/plugins/Widgets.jsx (hunk 3 above)
Comment (alex-fko) [reply web/client/plugins/Widgets.jsx:254]: It doesn't make sense to allow to disable select list completely, navigation between widgets is not possible without it. While arrows is an additional navigation method (to be honest I prefer it more than a select list) that could be disabled to free up some space for select list. Maybe we should allow to toggle between these two navigation methods or use both of them?

File: web/client/themes/default/less/widget.less (hunk 4 above)
Comment (alex-fko) [reply web/client/themes/default/less/widget.less:205]: This is done to make it's possible to have select list dropdown as wide as its possible without breaking its default behavior. If I limit select width, internal menu element is limited by this width, making dropdown too narrow to include long words without a breaks.
Arrows with transform:
Screenshot_31

Without transform:
Screenshot_32

TLDR; to free up some space for dropdown menu

File: web/client/themes/default/less/widget.less (hunk 5 above)
Comment (alex-fko) [reply web/client/themes/default/less/widget.less:191]: At least for the height, yes, widget height is 28px while react-select default style is 34px and in this version of react-select that's the only way to control its height (by rewriting it in a few places). Rest of styles are for its proper work with arrow navigation including `transform` property.

**Comment by alex-fko** [comment]
@allyoucanmap Please re-review. As discussed during the call:
- Code is simplified and now activates single widget view based on window width OR if there is a mobile browser user agent.
- Naming changes;
- Simplified styling for single widget layout, widget header.

Regarding grid layout changing from multi-column to single-column and back to multi-column:
- Grid has breakpoints that are activated based on grid width (not viewport width).
- There was a condition like "if screen width is > 800px, then add an offset 500px from the left" and this effectively was shrinking available space for widgets down to 300px when viewport is 801px.
- So 0-600px width: 1 column layout based on viewport width; 600-800px - 4 column layout based on grid width; 801-960px - 1 column due to grid width <480px; 960px and up - 4 columns.

My concern is about the changes related to this. I simplified this condition to not use left offset at all, regardless of the screen width, it removed unwanted layout changes, but now widgets can be overlapped by editor panel. But that was the case previously as well, for screen sizes with width 800px and less widgets were overlapped in some cases. We can minimize this behavior by checking if left panel is open and dynamically change grid width by adding offset from the left ONLY if it won't hit the breakpoint that changes layout to single column. @allyoucanmap what do you think?

File: web/client/components/widgets/widget/SingleWidgetControls.jsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import Button from "../../../components/misc/Button";
+
+const SingleWidgetControls = ({ options }) => {
+    const maxIndex = options.dropdownWidgets.length - 1;
+    const activeIndex = options.dropdownWidgets.findIndex(el => el.id === options.activeWidget.id);
+    const nextIndex = activeIndex === maxIndex ? 0 : activeIndex + 1;
+    const prevIndex = activeIndex === 0 ? maxIndex : activeIndex - 1;
+    return (
+        <div className="widget-selector">
+            {options.showArrowsNavigation ?
```
Comment (allyoucanmap) [inline web/client/components/widgets/widget/SingleWidgetControls.jsx:15]: please remove `options.showArrowsNavigation` is not requested in the issue

File: web/client/components/widgets/widget/SingleWidgetControls.jsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                    className={classnames({
+                        "Select-option": true,
+                        "is-focused": option.active
+                    })}
+                    onPointerDown={(e) => onSelect(option, e)}
+                >
+                    {option.label}
+                </div>}
+            />
+            {options.showArrowsNavigation ?
```
Comment (allyoucanmap) [inline web/client/components/widgets/widget/SingleWidgetControls.jsx:43]: please remove options.showArrowsNavigation is not requested in the issue

File: web/client/plugins/Widgets.jsx [hunk 8]
```diff
@@ -60,10 +81,16 @@ compose(
     compose(
         heightProvider({ debounceTime: 20, closest: true, querySelector: '.fill' }),
         widthProvider({ overrideWidthProvider: false }),
-        withProps(({width, height, maximized, mapLayout} = {}) => {
+        withProps(({ isMobileAgent }) => {
```
Comment (allyoucanmap) [inline web/client/plugins/Widgets.jsx:84]: please try to apply this patch:

- review of width/height breakpoints
- move the singleWidgetLayoutBreakpoint to a prop

[resize.zip](https://github.com/geosolutions-it/MapStore2/files/9703862/resize.zip)

File: web/client/plugins/Widgets.jsx [hunk 9]
```diff
@@ -160,6 +242,7 @@ class Widgets extends React.Component {
  * @memberof plugins
  * @name Widgets
  * @class
+ * @prop {boolean} [showArrowsNavigation] show arrows to toggle between available widgets in mobile view. True by default.
```
Comment (allyoucanmap) [inline web/client/plugins/Widgets.jsx:245]: please remove this cfg

File: web/client/plugins/widgets/WidgetsTray.jsx [hunk 10]
```diff
@@ -23,7 +23,7 @@ import BButton from '../../components/misc/Button';
 import {mapLayoutValuesSelector} from "../../selectors/maplayout";
 
 const Button = tooltip(BButton);
-
+const singleWidgetLayoutBreakpoint = 600;
```
Comment (allyoucanmap) [inline web/client/plugins/widgets/WidgetsTray.jsx:26]: provide this breakpoint as a prop with default value 600.

File: web/client/plugins/widgets/WidgetsTray.jsx [hunk 11]
```diff
@@ -80,14 +80,17 @@ class WidgetsTray extends React.Component {
...
     };
     static defaultProps = {
         enabled: true,
         items: [],
         expanded: false,
         setExpanded: () => { },
-        layout: {}
+        layout: {},
+        isMobile: false,
+        isTablet: false
```
Comment (allyoucanmap) [inline web/client/plugins/widgets/WidgetsTray.jsx:93]: please remove this default props

File: web/client/plugins/widgets/WidgetsTray.jsx [hunk 12]
```diff
@@ -126,6 +134,11 @@ export default compose(
         hasCollapsedWidgets: widgets.filter(({ collapsed } = {}) => collapsed).length > 0,
         hasTrayWidgets: widgets.length > 0
     })),
+    withProps(({ isMobileAgent }) => {
+        return {
+            isSingleWidgetLayout: isMobileAgent || window.innerWidth <= singleWidgetLayoutBreakpoint
+        };
```
Comment (allyoucanmap) [inline web/client/plugins/widgets/WidgetsTray.jsx:140]: this should have a similar logic we applied in the Widgets plugin, see the resize.patch. We should avoid to use  window.innerWidth and use a width detector (eg the HOC `import { withResizeDetector } from 'react-resize-detector';` )

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

File: web/client/plugins/Widgets.jsx (hunk 3 above)
Comment (allyoucanmap) [reply web/client/plugins/Widgets.jsx]: I would prefer to remove this option for the moment as mentioned in the next review

File: web/client/plugins/Widgets.jsx (hunk 9 above)
Comment (alex-fko) [reply web/client/plugins/Widgets.jsx]: removed

File: web/client/plugins/Widgets.jsx (hunk 8 above)
Comment (alex-fko) [reply web/client/plugins/Widgets.jsx]: thanks, applied the patch

File: web/client/components/widgets/widget/SingleWidgetControls.jsx (hunk 7 above)
Comment (alex-fko) [reply web/client/components/widgets/widget/SingleWidgetControls.jsx]: removed

File: web/client/components/widgets/widget/SingleWidgetControls.jsx (hunk 6 above)
Comment (alex-fko) [reply web/client/components/widgets/widget/SingleWidgetControls.jsx]: removed

File: web/client/plugins/widgets/WidgetsTray.jsx (hunk 10 above)
Comment (alex-fko) [reply web/client/plugins/widgets/WidgetsTray.jsx:27]: provided via the patch you've posted, thanks

File: web/client/plugins/widgets/WidgetsTray.jsx (hunk 12 above)
Comment (alex-fko) [reply web/client/plugins/widgets/WidgetsTray.jsx]: Unfortunately react-resize-detector didn't work for WidgetsTray (undefined width and height), while react-container-dimensions as a HOC/enhancer worked well.

**Comment by ElenaGallo** [comment]
@alex-fko a few notes after testing the issue:

- If there is only one widget the arrows to move through the widgets are enabled

*How to Reproduce*
[Example map](https://dev-mapstore.geosolutionsgroup.com/mapstore/#/viewer/openlayers/41726)

https://user-images.githubusercontent.com/56537133/194564389-2dca8eb2-7215-494a-975a-fbdf5ed1afa5.mov

- If there are more than one widget the collapse button does not work as expected

*How to Reproduce*
[Example map](https://dev-mapstore.geosolutionsgroup.com/mapstore/#/viewer/openlayers/10358)

https://user-images.githubusercontent.com/56537133/194564791-a7eb1b7f-a852-4fbf-b774-f9a14e8977f9.mov

**Comment by alex-fko** [comment]
Thanks @ElenaGallo, I'll raise PR with fixes for these two problems asap.
